*: support bazel

hawkingrei opened the pull request:
<!--

Thank you for contributing to TiDB!

PR Title Format:
1. pkg [, pkg2, pkg3]: what's changed
2. *: what's changed

-->

### What problem does this PR solve?
<!--

Please create an issue first to describe the problem.

There MUST be one line starting with "Issue Number:  " and 
linking the relevant issues via the "close" or "ref".

For more info, check https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/contribute-code.html#referring-to-an-issue.

-->

Issue Number: close #34083

Problem Summary:

### What is changed and how it works?

### Check List

Tests <!-- At least one of them must be included. -->

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

<!-- bugfix or new feature needs a release note -->

```release-note
None
```

zz-jason commented:
could you explain more about why adding this support?

hawkingrei commented:
> could you explain more about why adding this support?

We run thousands of builds and tests in a day. But I find that we have any local cache and remote cache to speed up ci. so  I import the bazel into tidb to solve this problem. I believe that it can boost performance 10x.

hawkingrei commented:
> could you explain more about why adding this support?

if I meet an unstable test in the CI, I will just rerun this test case by the bazel. 
if I have a test case that it just tests ```util``` package and ```util``` package have not any changes, we can use the remote cache to skip this test.
![image](https://user-images.githubusercontent.com/3427324/163121483-41945dfd-43ac-45d4-a54f-e3c01ea71cc2.png)


image.png

breeswish commented:
How about using a local cache that is provided by Go natively? For example, modern CI like GitHub Action provides per-branch cache, which is sufficient to be used as a cache, even incremental one.

I made a demo of utilizing GitHub Action cache:

1. [Full test run](https://github.com/breeswish/go-test-cache-demo/runs/6038147567?check_suite_focus=true)

   ```
   ok  	github.com/breeswish/go-test-cache-demo/bar	222.714s
   ok  	github.com/breeswish/go-test-cache-demo/box	22.670s
   ok  	github.com/breeswish/go-test-cache-demo/foo	222.729s
   ```

2. [With a cache, update a single package](https://github.com/breeswish/go-test-cache-demo/runs/6038730204?check_suite_focus=true)

   ```
   ok  	github.com/breeswish/go-test-cache-demo/bar	220.462s
   ok  	github.com/breeswish/go-test-cache-demo/box	(cached)
   ok  	github.com/breeswish/go-test-cache-demo/foo	(cached)
   ```

3. [In the same PR, make change to non-code](https://github.com/breeswish/go-test-cache-demo/runs/6038894438?check_suite_focus=true): You can see the cache from 2 is effective, so that CI is finished immediately, only take seconds

   ```
   ok  	github.com/breeswish/go-test-cache-demo/bar	(cached)
   ok  	github.com/breeswish/go-test-cache-demo/box	(cached)
   ok  	github.com/breeswish/go-test-cache-demo/foo	(cached)
   ```

We can use the same strategy in our own CI infrastructure to utilize the same effect:

1. For master commit: Test and save cache
2. For PR commit: Read a recent cache from current PR (if exists), or a recent cache from master

In this way, PR test can be fast, when the PR is first proposed, or the PR is updated incrementally.

hawkingrei commented:
> How about using a local cache that is provided by Go natively? For example, modern CI like GitHub Action provides per-branch cache, which is sufficient to be used as a cache, even incremental one.
> 
> I made a demo of utilizing GitHub Action cache:
> 
>     1. [Full test run](https://github.com/breeswish/go-test-cache-demo/runs/6038147567?check_suite_focus=true)
>        ```
>        ok  	github.com/breeswish/go-test-cache-demo/bar	222.714s
>        ok  	github.com/breeswish/go-test-cache-demo/box	22.670s
>        ok  	github.com/breeswish/go-test-cache-demo/foo	222.729s
>        ```
> 
>     2. [With a cache, update a single package](https://github.com/breeswish/go-test-cache-demo/runs/6038730204?check_suite_focus=true) ([commit](https://github.com/breeswish/go-test-cache-demo/pull/5/commits/0ef0e72effcf6f3fbe620149b0e817d88b5d359e)): You can see that test results of untouched packages are reused (from the master branch)
>        ```
>        ok  	github.com/breeswish/go-test-cache-demo/bar	220.462s
>        ok  	github.com/breeswish/go-test-cache-demo/box	(cached)
>        ok  	github.com/breeswish/go-test-cache-demo/foo	(cached)
>        ```
> 
>     3. [In the same PR, make change to non-code](https://github.com/breeswish/go-test-cache-demo/runs/6038894438?check_suite_focus=true) ([commit](https://github.com/breeswish/go-test-cache-demo/pull/5/commits/a6a20405b9c89f91dec6d8f62bb203f1a4fd213f)): You can see the cache from 2 is effective, so that CI is finished immediately, only take seconds
>        ```
>        ok  	github.com/breeswish/go-test-cache-demo/bar	(cached)
>        ok  	github.com/breeswish/go-test-cache-demo/box	(cached)
>        ok  	github.com/breeswish/go-test-cache-demo/foo	(cached)
>        ```
> 
> 
> We can use the same strategy in our own CI infrastructure to utilize the same effect:
> 
>     1. For master commit: Test and save cache
> 
>     2. For PR commit: Read a recent cache from current PR (if exists), or a recent cache from master
> 
> 
> In this way, PR test can be fast, when the PR is first proposed, or the PR is updated incrementally.

if you have a remote cache, it can do better.

hawkingrei commented:
> How about using a local cache that is provided by Go natively? For example, modern CI like GitHub Action provides per-branch cache, which is sufficient to be used as a cache, even incremental one.
> 
> I made a demo of utilizing GitHub Action cache:
> 
>     1. [Full test run](https://github.com/breeswish/go-test-cache-demo/runs/6038147567?check_suite_focus=true)
>        ```
>        ok  	github.com/breeswish/go-test-cache-demo/bar	222.714s
>        ok  	github.com/breeswish/go-test-cache-demo/box	22.670s
>        ok  	github.com/breeswish/go-test-cache-demo/foo	222.729s
>        ```
> 
>     2. [With a cache, update a single package](https://github.com/breeswish/go-test-cache-demo/runs/6038730204?check_suite_focus=true) ([commit](https://github.com/breeswish/go-test-cache-demo/pull/5/commits/0ef0e72effcf6f3fbe620149b0e817d88b5d359e)): You can see that test results of untouched packages are reused (from the master branch)
>        ```
>        ok  	github.com/breeswish/go-test-cache-demo/bar	220.462s
>        ok  	github.com/breeswish/go-test-cache-demo/box	(cached)
>        ok  	github.com/breeswish/go-test-cache-demo/foo	(cached)
>        ```
> 
>     3. [In the same PR, make change to non-code](https://github.com/breeswish/go-test-cache-demo/runs/6038894438?check_suite_focus=true) ([commit](https://github.com/breeswish/go-test-cache-demo/pull/5/commits/a6a20405b9c89f91dec6d8f62bb203f1a4fd213f)): You can see the cache from 2 is effective, so that CI is finished immediately, only take seconds
>        ```
>        ok  	github.com/breeswish/go-test-cache-demo/bar	(cached)
>        ok  	github.com/breeswish/go-test-cache-demo/box	(cached)
>        ok  	github.com/breeswish/go-test-cache-demo/foo	(cached)
>        ```
> 
> 
> We can use the same strategy in our own CI infrastructure to utilize the same effect:
> 
>     1. For master commit: Test and save cache
> 
>     2. For PR commit: Read a recent cache from current PR (if exists), or a recent cache from master
> 
> 
> In this way, PR test can be fast, when the PR is first proposed, or the PR is updated incrementally.

Why we cannot use the golang 

> How about using a local cache that is provided by Go natively? For example, modern CI like GitHub Action provides per-branch cache, which is sufficient to be used as a cache, even incremental one.
> 
> I made a demo of utilizing GitHub Action cache:
> 
>     1. [Full test run](https://github.com/breeswish/go-test-cache-demo/runs/6038147567?check_suite_focus=true)
>        ```
>        ok  	github.com/breeswish/go-test-cache-demo/bar	222.714s
>        ok  	github.com/breeswish/go-test-cache-demo/box	22.670s
>        ok  	github.com/breeswish/go-test-cache-demo/foo	222.729s
>        ```
> 
>     2. [With a cache, update a single package](https://github.com/breeswish/go-test-cache-demo/runs/6038730204?check_suite_focus=true) ([commit](https://github.com/breeswish/go-test-cache-demo/pull/5/commits/0ef0e72effcf6f3fbe620149b0e817d88b5d359e)): You can see that test results of untouched packages are reused (from the master branch)
>        ```
>        ok  	github.com/breeswish/go-test-cache-demo/bar	220.462s
>        ok  	github.com/breeswish/go-test-cache-demo/box	(cached)
>        ok  	github.com/breeswish/go-test-cache-demo/foo	(cached)
>        ```
> 
>     3. [In the same PR, make change to non-code](https://github.com/breeswish/go-test-cache-demo/runs/6038894438?check_suite_focus=true) ([commit](https://github.com/breeswish/go-test-cache-demo/pull/5/commits/a6a20405b9c89f91dec6d8f62bb203f1a4fd213f)): You can see the cache from 2 is effective, so that CI is finished immediately, only take seconds
>        ```
>        ok  	github.com/breeswish/go-test-cache-demo/bar	(cached)
>        ok  	github.com/breeswish/go-test-cache-demo/box	(cached)
>        ok  	github.com/breeswish/go-test-cache-demo/foo	(cached)
>        ```
> 
> 
> We can use the same strategy in our own CI infrastructure to utilize the same effect:
> 
>     1. For master commit: Test and save cache
> 
>     2. For PR commit: Read a recent cache from current PR (if exists), or a recent cache from master
> 
> 
> In this way, PR test can be fast, when the PR is first proposed, or the PR is updated incrementally.

The size of TiDB cache which includes all builds and tests is 19GB. it is too big to upload all cache to S3 and download all cache the next time. and we cannot control the job task to make it run the same machine and use cache every time. it is so hard because Jenkins is not k8s CRD. we cannot schedule anything. so TiDB CI does not have any cache.

hawkingrei commented:
Bazel has a better design. it can use remote cache to achieve distributed compilation and testing.

hawkingrei commented:
/check-issue-triage-complete

winoros reviewed (commented):
I just have a simple question.
When we updating the `go.mod`, how we make sure the dependencies in `bzl` is up to date?

hawkingrei commented:
> I just have a simple question. When we updating the `go.mod`, how we make sure the dependencies in `bzl` is up to date?

use a robot to verify it.

tiancaiamao commented:
Can we avoid adding the BUILD.bazel and put the generating process in the script?
 
```
bazel run //:gazelle
```

hawkingrei commented:
> Can we avoid adding the BUILD.bazel and put the generating process in the script?
> 
> ```
> bazel run //:gazelle
> ```

Yes, We can do it.

breeswish commented:
> The size of TiDB cache which includes all builds and tests is 19GB. it is too big to upload all cache to S3 and download all cache the next time. and we cannot control the job task to make it run on the same machine and use local cache every time. it is hard because Jenkins is not k8s CRD. we cannot schedule anything. so TiDB CI does not have any cache now.

This is a good point! Previously in Jenkins CI we usually use local cache. Network cache is too expansive. Maybe you can ask EE for help. I think caching is a totally valid requirement.

hawkingrei commented:
> Can we avoid adding the BUILD.bazel and put the generating process in the script?
> 
> ```
> bazel run //:gazelle
> ```

Done

tiancaiamao commented:
It's a hard decision to make. I'd say, I like the potential performance boost by this change.
However, it comes with costs: how do we maintain this toolchain in the future?
Bazel has its learning curve, it's difficult for find someone else who is familiar with bazel within the team.

Also, mentioned here https://github.com/golang/go/issues/42785#issuecomment-740017077

>  it's detrimental for builds

It's better just to use it for CI test environment, we still use `go` to build the release binary.

hawkingrei commented:
> It's a hard decision to make. I'd say, I like the potential performance boost by this change. However, it comes with costs: how do we maintain this toolchain in the future? Bazel has its learning curve, it's difficult for find someone else who is familiar with bazel within the team.
> 
> Also, mentioned here [golang/go#42785 (comment)](https://github.com/golang/go/issues/42785#issuecomment-740017077)
> 
> > it's detrimental for builds
> 
> It's better just to use it for CI test environment, we still use `go` to build the release binary.

bazel just is a tool that can manage the build/test cache and execute distributed builds and tests. His configuration can be automatically generated. It is easy to use. so we still can ```go``` to build the release binary.

tisonkun commented on config/config_test.go:
```diff
@@ -496,8 +495,6 @@ func TestConflictInstanceConfig(t *testing.T) {
 	var expectedNewName string
 	conf := new(Config)
 	configFile := "config.toml"
-	_, localFile, _, _ := runtime.Caller(0)
```
Why make these changes?

hawkingrei replied:
It is useless code. but i think I can create a pr to merge it.

tisonkun commented on ddl/BUILD.bazel:
```diff
@@ -0,0 +1,229 @@
+go_library(
+    name = "ddl",
+    srcs = [
+        "backfilling.go",
```
Shall we manually update this list everytime we add, delete or rename files?

hawkingrei replied:
No, This job is automatic.

tisonkun replied:
Thanks for your information. You may create an issue on https://github.com/pingcap/tidb-dev-guide later and add some words about working with bazel.

hawkingrei replied:
[a good examle](https://wiki.crdb.io/wiki/spaces/CRDB/pages/2221703221/Developing+with+Bazel)

tisonkun commented on executor/show_stats_test.go:
```diff
@@ -342,18 +342,6 @@ func TestShowColumnStatsUsage(t *testing.T) {
 	require.Equal(t, rows[1], []interface{}{"test", "t2", p0.Name.O, t1.Meta().Columns[0].Name.O, "2021-10-20 09:00:00", "<nil>"})
 }
 
-func TestShowHistogramsInFlight(t *testing.T) {
```
Why remove this test?

hawkingrei replied:
Bazel isolated the test by the folder. if too many tests are in the same folder, it will have many problems. so we need to move it into another package in a different folder.

tisonkun replied:
> many problems

Could you elaborate a bit on this statement?

hawkingrei replied:
I don't know why it happens. but I try to remove it into another folder to keep the tests independent. it is successful.

tisonkun replied:
Fine. Then let's remain as is.

tisonkun commented on session/session_test/BUILD.bazel:
```diff
@@ -0,0 +1,26 @@
+load("@io_bazel_rules_go//go:def.bzl", "go_test")
+
+go_test(
+    name = "session_test_test",
```
Maybe you can make a dedicated name. package session_test is ambitious.

hawkingrei replied:
OK, I agree with you.

tisonkun commented on tests/realtikvtest/testkit.go:
```diff
@@ -53,6 +53,7 @@ func RunTestMain(m *testing.M) {
 	})
 	tikv.EnableFailpoints()
 	opts := []goleak.Option{
+		goleak.IgnoreTopFunction("github.com/pingcap/goleveldb/leveldb.(*DB).mpoolDrain"),
```
Since we wait for goleveldb closed below, perhaps we don't need this line.

hawkingrei replied:
removed it.

tisonkun reviewed (commented):
util/testleak/BUILD.bazel should be removed? Since the package itself is dropped.

Other comments inline.

hawkingrei commented on BUILD.bazel:
```diff
@@ -0,0 +1,28 @@
+load("@bazel_gazelle//:def.bzl", "gazelle")
+
+# gazelle:proto disable_global
```
it is control code for build generate tool. so the human being can ignore it.

hawkingrei commented on Makefile:
```diff
@@ -429,3 +429,26 @@ tools/bin/gotestsum: tools/check/go.mod
 	@cd metrics/grafana && mv tidb_summary.json tidb_summary.json.committed && ./generate_json.sh && diff -u tidb_summary.json.committed tidb_summary.json && rm tidb_summary.json.committed
+
+bazel_prepare:
+	bazel --output_user_root=/home/jenkins/.tidb/tmp run --config=ci  //:gazelle
```
Yes, I will separate the dev env and CI env. Now it is only used for CI env. 

BTW, I don't want to change the developer's environment too much. The contributor doesn't have to install Bazel to develop tidb and needn't run anything to meet bazel.

hawkingrei replied:
The contributor shouldn't run it.

tisonkun reviewed (approved):
This change looks good to me. However, could you elaborate the following works in #34083?

This is a fundamental build system change and it deserves a design doc for the whole lifecycle and how other contributors should work with it.

hawkingrei commented:
> This change looks good to me. However, could you elaborate the following works in #34083?
> 
> This is a fundamental build system change and it deserves a design doc for the whole lifecycle and how other contributors should work with it.

it will be done, We still fight with Jenkins ci. It is so hard.

breezewish commented:
Just curious, if these BAZEL files are automatically generated, can they be git-ignored?

hawkingrei commented:
> Just curious, if these BAZEL files are automatically generated, can they be git-ignored? It's usually not a good practice to include generated files in the git repository.

In fact, Some bazel files are automatically generated. for example flaky test (It isn't recommended) and some other config have to be controlled by development such as visibility and test size that controls the test timeout. so it must be with the repo. 

BTW, the open resource repos using the bazel like [abseil-cpp](https://github.com/abseil/abseil-cpp), [cockroach](https://github.com/cockroachdb/cockroach), [tensorflow](https://github.com/tensorflow/tensorflow) have many BAZEl.build files in their repo.

breezewish commented:
> > Just curious, if these BAZEL files are automatically generated, can they be git-ignored? It's usually not a good practice to include generated files in the git repository.
> 
> In fact, Some bazel files are automatically generated. for example flaky test (It isn't recommended) and some other config have to be controlled by development such as visibility and test size that controls the test timeout. so it must be with the repo.
> 
> BTW, the open resource repos using the bazel like [abseil-cpp](https://github.com/abseil/abseil-cpp), [cockroach](https://github.com/cockroachdb/cockroach), [tensorflow](https://github.com/tensorflow/tensorflow) have many BAZEl.build files in their repo.

@hawkingrei Does this mean that we need to manually maintain these BAZEL.build files instead of using auto-generated ones, as the auto-generated ones may not be desired? I would like to know, for example, when I introduced a new package in the TiDB repository, what steps should I take in order to make sure everything works?

breezewish commented:
Maybe you also need to introduce a CI step to ensure that people doesn't **forget** to generate necessary BAZEL files? I'm not sure however.

hawkingrei commented:
> Maybe you also need to introduce a CI step to ensure that people doesn't **forget** to generate necessary BAZEL files? I'm not sure however.

In fact, Generate bazel config is a task that takes some time in the CI. Althought it take less time than that tests without cache. so I will remove it in the future to speed ci again in the future. 

However, Gezalle just can manage the dependency in the bazel config. it cannot manage the other thing. if we need to manage the config such as flasky test and so on. Gazalle just copy the old config to the new config. it is the main reason why we cannot remove ```bazel.build```.

hawkingrei commented:
> > > Just curious, if these BAZEL files are automatically generated, can they be git-ignored? It's usually not a good practice to include generated files in the git repository.
> > 
> > 
> > In fact, Some bazel files are automatically generated. for example flaky test (It isn't recommended) and some other config have to be controlled by development such as visibility and test size that controls the test timeout. so it must be with the repo.
> > BTW, the open resource repos using the bazel like [abseil-cpp](https://github.com/abseil/abseil-cpp), [cockroach](https://github.com/cockroachdb/cockroach), [tensorflow](https://github.com/tensorflow/tensorflow) have many BAZEl.build files in their repo.
> 
> @hawkingrei Does this mean that we need to manually maintain these BAZEL.build files instead of using auto-generated ones, as the auto-generated ones may not be desired? I would like to know, for example, when I introduced a new package in the TiDB repository (or simply, added a new file), what steps should I take in order to make sure everything works? How these BAZEL files need to be generated? What kind of modification should I make? etc. If these steps are not property done, will my tests just being ignored in CI?

For the common developers, It will be fine that they use gazelle to generate the config. For the repo admin, it needs to control the other thing by changing the config in the bazel. in these cases, we need to change the config and save it, so that gazelle can get this Changement and keep them in the new config if the dependency has been updated.

bb7133 reviewed (approved):
LGTM

I had a few questions which are the same from @tison, and I agree with the feedback, at least for now.

bb7133 commented:
/merge
